### PR TITLE
Watch Mode and Build Enhancements

### PR DESCRIPTION
**Pull request by blink1073**
- Automatically watch local TypeScript files for changes when using `jupyter lab --dev-mode --watch`.
- Streamline the test watch mode
- Add logic for when linked packages have changed
- Only build if necessary in app dir watch mode
- Do not run client build check in core mode
- Add configuration option to skip automatic build check but still enable the build tool

Bottom line for watch mode (dev, app, test):
- Make sure we have a valid build before starting
- Watch only the files we care about
- If necessary, add a shim to move files to where webpack can see them

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "test:ie": "lerna run test:ie --concurrency 1 --stream",
     "update:dependency": "node scripts/update-dependency.js",
     "update:core": "cd jupyterlab && node update-core.js",
-    "watch": "watch \"npm run build:packages\" ./packages/** --wait 10 --filter=scripts/watch-filter.js --ignoreDotFiles",
```
Comment (ian-r-rose) [inline package.json]: Can watch-filter.js be removed entirely now?

File: packages/all-packages/package.json [hunk 2]
```diff
@@ -27,7 +27,7 @@
     "build": "tsc && node build.js",
     "clean": "rimraf lib",
     "docs": "typedoc --mode modules --module commonjs --excludeNotExported --target es5 --moduleResolution node --name JupyterLab --out ../../docs/api .",
-    "watch": "watch \"npm run build\" .. --wait 10 --filter=../../scripts/watch-filter.js"
+    "watch": "tsc-watch --onSuccess \"node build.js\""
```
Comment (ian-r-rose) [inline packages/all-packages/package.json]: Can we continue to ignore dotfiles for vi users such as myself?

Comment (blink1073) [reply packages/all-packages/package.json]: TypeScript only watches the TS files in your project.

File: package.json (hunk 1 above)
Comment (blink1073) [reply package.json]: Removed now, we were also using it in `test/`.

File: packages/all-packages/package.json (hunk 2 above)
Comment (ian-r-rose) [reply packages/all-packages/package.json]: I get an error running `tsc-watch` due to a `\r`: 
```bash
/usr/bin/env: ‘node\r’: No such file or directory
```

Comment (blink1073) [reply packages/all-packages/package.json]: Looks like we need to pin the version: https://github.com/gilamran/tsc-watch/issues/6

Comment (ian-r-rose) [reply packages/all-packages/package.json]: I can confirm that it works with `tsc-watch@1.0.7`.

Puzzlingly, it looks like it is recompiling twice upon file changes? At least it is logging to output twice:
```
7:45:28 PM - File change detected. Starting incremental compilation...
7:45:28 PM - File change detected. Starting incremental compilation...
7:45:32 PM - Compilation complete. Watching for file changes.
7:45:32 PM - Compilation complete. Watching for file changes.
Hash: fa35d1f5beb69f09c741
Version: webpack 2.7.0
Time: 1253ms
             Asset     Size  Chunks                    Chunk Names
    main.bundle.js  7.67 MB       1  [emitted]  [big]  main
main.bundle.js.map  9.17 MB       1  [emitted]         main
   [6] ../packages/coreutils/lib/index.js 870 bytes {1} [built]
  [27] ../packages/application/lib/index.js 5.64 kB {1} [built]
 [451] ../packages/imageviewer-extension/lib/index.js 4.11 kB {1} [optional] [built]
 [454] ../packages/markdownviewer-extension/lib/index.js 2.37 kB {1} [optional] [built]
 [455] ../packages/notebook-extension/lib/index.js 39.8 kB {1} [optional] [built]
 [456] ../packages/pdf-extension/lib/index.js 4.96 kB {1} [optional] [built]
 [457] ../packages/running-extension/lib/index.js 1.67 kB {1} [optional] [built]
 [458] ../packages/settingeditor-extension/lib/index.js 2.63 kB {1} [optional] [built]
 [459] ../packages/shortcuts-extension/lib/index.js 3.71 kB {1} [optional] [built]
 [460] ../packages/tabmanager-extension/lib/index.js 1.77 kB {1} [optional] [built]
 [461] ../packages/terminal-extension/lib/index.js 7.3 kB {1} [optional] [built]
 [462] ../packages/theme-dark-extension/lib/index.js 767 bytes {1} [optional] [built]
 [463] ../packages/theme-light-extension/lib/index.js 771 bytes {1} [optional] [built]
 [464] ../packages/tooltip-extension/lib/index.js 5.57 kB {1} [optional] [built]
 [465] ../packages/vega2-extension/lib/index.js 6.17 kB {1} [optional] [built]
    + 1147 hidden modules
Hash: fa35d1f5beb69f09c741
Version: webpack 2.7.0
Time: 1256ms
             Asset     Size  Chunks                    Chunk Names
    main.bundle.js  7.67 MB       1  [emitted]  [big]  main
main.bundle.js.map  9.17 MB       1  [emitted]         main
   [6] ../packages/coreutils/lib/index.js 870 bytes {1} [built]
  [27] ../packages/application/lib/index.js 5.64 kB {1} [built]
 [451] ../packages/imageviewer-extension/lib/index.js 4.11 kB {1} [optional] [built]
 [454] ../packages/markdownviewer-extension/lib/index.js 2.37 kB {1} [optional] [built]
 [455] ../packages/notebook-extension/lib/index.js 39.8 kB {1} [optional] [built]
 [456] ../packages/pdf-extension/lib/index.js 4.96 kB {1} [optional] [built]
 [457] ../packages/running-extension/lib/index.js 1.67 kB {1} [optional] [built]
 [458] ../packages/settingeditor-extension/lib/index.js 2.63 kB {1} [optional] [built]
 [459] ../packages/shortcuts-extension/lib/index.js 3.71 kB {1} [optional] [built]
 [460] ../packages/tabmanager-extension/lib/index.js 1.77 kB {1} [optional] [built]
 [461] ../packages/terminal-extension/lib/index.js 7.3 kB {1} [optional] [built]
 [462] ../packages/theme-dark-extension/lib/index.js 767 bytes {1} [optional] [built]
 [463] ../packages/theme-light-extension/lib/index.js 771 bytes {1} [optional] [built]
 [464] ../packages/tooltip-extension/lib/index.js 5.57 kB {1} [optional] [built]
 [465] ../packages/vega2-extension/lib/index.js 6.17 kB {1} [optional] [built]
    + 1147 hidden modules
```

Comment (blink1073) [reply packages/all-packages/package.json]: Switched to using the `watch` library, which only copies and updates the webpack bundle for files that have actually changed.

**Comment by blink1073** [comment]
Needs a bit more tweaking to not copy all the files on the first build.

**Comment by ian-r-rose** [comment]
Looking good. Using `watch` is indeed moderately faster for me.

**Comment by gnestor** [comment]
So much better now! Thanks @blink1073!!

**Pull request merged**